### DISCUSSION TITLE
Injecting value into either a custom discovery provider - either provider or service?

### DISCUSSION THREAD

**Discussion by kjq** [smallrye/smallrye-stork#608] [Q&A] [answered]
Quick, hopefully simple, question.

I have a custom provider, based off of the Kubernetes discovery provider, that is being used in Quarkus - I would like to inject some custom values into either but I do not seem capable of doing it.

It does not fail on the annotation `@ConfigProperty` but the value is always null.

Am I missing something?

**Comment by cescoffier** [comment]
Hello, 

In Quarkus 3, the service discovery providers are cdi beans and so can use @Inject to retrieve config value. Note that only the provider is, not the service discovery itself.

For Quarkus 2, you would need to use the CDI API (CDI.current().select(...)).

**Reply by kjq** [reply]
Odd, I felt dumb asking because that is what I thought but at least when I use `@ConfigProperty` it does not seem to work.  Ill try injecting something else and check again just to make sure.

**Reply by kjq** [reply]
Weird, `@Inject` value is always null when used inside my discovery  or provider implementation.  I dont think it is a bug just something I am doing wrong, maybe with injecting Stork using the `@Provider` in a config?

Attached is a crude reproducer with a couple of System.out's.  In both cases the injected properties are null.  I just put a breakpoint in each of them and inspected the injected value.

Other than that everything seems to be working as its supposed to.

[code-with-quarkus.zip](https://github.com/smallrye/smallrye-stork/files/11962562/code-with-quarkus.zip)

**Reply by kjq** [reply]
I just realized what you said about the discovery _not_ being a CDI bean.  Still, my provider is not recognizing anything i inject into it.

**Reply by cescoffier** [reply, marked as answer]
To be a bean, the providers must be annotated with `@ApplicationScoped` (see http://smallrye.io/smallrye-stork/2.3.0/service-discovery/custom-service-discovery/#implementing-a-service-discovery-provider).

Then, pass the bean to the `ServiceDiscovery` implementation constructor.

**Reply by kjq** [reply]
Ugh, how did I miss that?  I think I would look at the generated ProviderLoader and was assuming the provider was scoped.  Thanks!
